### DISCUSSION TITLE
Create a hole (circle) inside rectangle

### DISCUSSION THREAD

**Discussion by afwn90cj93201nixr2e1re** [ImageMagick/ImageMagick#3421] [Help] [answered]
.

**Comment by snibgo** [comment]
Your question is vague. You can draw rectangles and circles with the "-draw" operation. A circle can be made to the same size (width and height) as a square, but not a general rectangle.

An ellipse can be made the same size as a general rectangle.

What version of IM, on what platform?

**Comment by snibgo** [comment]
What version of IM, on what platform?

**Comment by snibgo** [comment]
Bash script, with IM v7:
```
myW=200
myH=200

magick -size ${myW}x${myH} xc:Black -fill White -draw 'circle %[fx:w/2],%[fx:h/2],%[fx:w/2],1' out.png
```

Better:
```
magick -size ${myW}x${myH} xc:Black -fill White -draw 'circle %[fx:w/2-0.5],%[fx:h/2-0.5],%[fx:w/2-0.5],0.5' out2.png
```

**Comment by snibgo** [comment]
I showed bash, because that is what you said. Personally, I mostly use Windows. For IM usage under Windows, see https://legacy.imagemagick.org/Usage/windows/ . Yes, Windows syntax is different to bash, and double-quotes are needed.

For Windows BAT scripts, double any percent signs that IM needs to see: %%.

> ... and if you can post 2 different things to make corners more rounded, like smoothed and non-smoothed way.

I think you mean antialiasing, where pixels at edges of shapes are blended rather than binary inside versus outside. Antialiasing is enabled by default. To turn it off, use "+antialias" before drawing the shape.

**Comment by snibgo** [comment, marked as answer]
Windows BAT scripts. First, with anti-alias ("smoothing"), the default. I crop and scale for clarity.
```
magick ^
  -size %myW%x%myH% xc:Black ^
  -fill White -draw "circle %%[fx:w/2-0.5],%%[fx:h/2-0.5],%%[fx:w/2-0.5],0.5" ^
  -crop 50x50+0+0 +repage ^
  -scale 400%% ^
  out2a.png
```
![out2a](https://user-images.githubusercontent.com/33812368/111824504-364c0100-88de-11eb-9ebe-8dd88a913929.png)

Now, "+antialias" to turn anti-aliasing off:
```
magick ^
  -size %myW%x%myH% xc:Black ^
  +antialias ^
  -fill White -draw "circle %%[fx:w/2-0.5],%%[fx:h/2-0.5],%%[fx:w/2-0.5],0.5" ^
  -crop 50x50+0+0 +repage ^
  -scale 400%% ^
  out2na.png
```
![out2na](https://user-images.githubusercontent.com/33812368/111824538-41069600-88de-11eb-96d5-3e582b9a1c3c.png)

For making transparent shapes, I would do as you did, make a black and whte clone (black where you want transparency), and use "-compose CopyOpacity -composite".

```
magick ^
  -size %myW%x%myH% xc:White ^
  (  +clone ^
    -fill Black -draw "circle %%[fx:w/2-0.5],%%[fx:h/2-0.5],%%[fx:w/2-0.5],0.5" ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  out2t.png
```
![out2t](https://user-images.githubusercontent.com/33812368/111824560-495ed100-88de-11eb-92ba-34257841ce2b.png)
The image is not easily visible on the forum.

**Reply by afwn90cj93201nixr2e1re** [reply]
```
@ECHO OFF
set myW=640
set myH=640

magick convert -size %myW%x%myH% xc:White (  +clone -fill Black -draw "circle %%[fx:w/2-0.5],%%[fx:h/2-0.5],%%[fx:w/2-0.5],0.5" ) -alpha off -compose CopyOpacity -composite out2t.png

pause
```

Also, are you sure about double percentage symbols in this case? Because it says: `convert: non-conforming drawing primitive definition `circle' @ error/draw.c/RenderMVGContent/4466.`.

**Comment by snibgo** [comment]
Why are you using "magick convert"? That gives you v6 syntax with v7 processing, and v6 syntax does't allow FX expressions within "-draw". Remove "convert".

**Comment by GeeMack** [comment]
There are many ways to create a circle mask and punch out a hole with ImageMagick. Some may be better than others for different workflows or to fit other parts of a command. Here's a method I use to punch out the largest possible full circle from the center of any image. This is in Windows syntax with IMv6...

    convert wizard: -alpha set -fill white -background none ^
       ( +clone -alpha off -colorize 100 -alpha set ^
             -splice 0x1 -distort arc "360 0 %[fx:min(w,h)/2-1] 0" ) ^
       -gravity center -compose dstout -composite wizardhole.png

That uses "-distort arc" to create a white circle on a transparent background, with a diameter of the smaller of the width or height of the input image. Then it uses a compose method of "dstout" so the composite more or less removes that white circle from the center of the input image. Change the gravity and/or set the geometry to put the "hole" in another location.

For IMv7 change "convert" to "magick".

On a *nix system you'll need to change the continued-line carets "^" to backslashes "\\" and escape the parentheses with backslashes "\\(...\\)".
